flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/factory/KubernetesJobManagerFactoryTest.java

```diff
@@ -0,0 +1,203 @@
+			this.kubernetesJobManagerSpecification.getDeployment().getSpec().getTemplate().getSpec();
+
+		assertEquals(1, resultPodSpec.getContainers().size());
+		assertEquals(SERVICE_ACCOUNT_NAME, resultPodSpec.getServiceAccountName());
+		assertEquals(1, resultPodSpec.getVolumes().size());
+
+		final Container resultedMainContainer = resultPodSpec.getContainers().get(0);
+		assertEquals(KubernetesJobManagerParameters.JOB_MANAGER_MAIN_CONTAINER_NAME, resultedMainContainer.getName());
+		assertEquals(CONTAINER_IMAGE, resultedMainContainer.getImage());
+		assertEquals(CONTAINER_IMAGE_PULL_POLICY, resultedMainContainer.getImagePullPolicy());
+
+		assertEquals(3, resultedMainContainer.getEnv().size());
+		assertTrue(resultedMainContainer.getEnv()
+				.stream()
+				.anyMatch(envVar -> envVar.getName().equals("key1")));
+
+		assertEquals(3, resultedMainContainer.getPorts().size());
+
+		final Map<String, Quantity> requests = resultedMainContainer.getResources().getRequests();
+		assertEquals(Double.toString(JOB_MANAGER_CPU), requests.get("cpu").getAmount());
+		assertEquals(JOB_MANAGER_MEMORY + "Mi", requests.get("memory").getAmount());
+
+		assertEquals(1, resultedMainContainer.getCommand().size());
+		assertEquals(3, resultedMainContainer.getArgs().size());
+
+		assertEquals(1, resultedMainContainer.getVolumeMounts().size());
+	}
+
+	@Test
+	public void testAdditionalResourcesSize() {
+		final List<HasMetadata> resultAdditionalResources = this.kubernetesJobManagerSpecification.getAccompanyingResources();
+		assertEquals(3, resultAdditionalResources.size());
+
+		final List<HasMetadata> resultServices = resultAdditionalResources
+			.stream()
+			.filter(x -> x instanceof Service)
+			.collect(Collectors.toList());
+		assertEquals(2, resultServices.size());
+
+		final List<HasMetadata> resultConfigMaps = resultAdditionalResources
+			.stream()
+			.filter(x -> x instanceof ConfigMap)
+			.collect(Collectors.toList());
+		assertEquals(1, resultConfigMaps.size());
+	}
+
+	@Test
+	public void testServices() {
+		final List<Service> resultServices = this.kubernetesJobManagerSpecification.getAccompanyingResources()
+			.stream()
+			.filter(x -> x instanceof Service)
+			.map(x -> (Service) x)
+			.collect(Collectors.toList());
+
+		assertEquals(2, resultServices.size());
+
+		final List<Service> internalServiceCandidates = resultServices
+				.stream()
+				.filter(x -> x.getMetadata().getName().equals(KubernetesUtils.getInternalServiceName(CLUSTER_ID)))
+				.collect(Collectors.toList());
+		assertEquals(1, internalServiceCandidates.size());
+
+		final List<Service> restServiceCandidates = resultServices
+				.stream()
+				.filter(x -> x.getMetadata().getName().equals(KubernetesUtils.getRestServiceName(CLUSTER_ID)))
+				.collect(Collectors.toList());
+		assertEquals(1, restServiceCandidates.size());
+
+		final Service resultInternalService = internalServiceCandidates.get(0);
+		assertEquals(2, resultInternalService.getMetadata().getLabels().size());
```

Comment:
Maybe we could use `getCommonLabels()` to assert the real labels, not only check the size. I think other `getLabels().size()` is in the same situation.

Reply:
Since we have checked the real labels in `InternalServiceDecoratorTest` and `ExternalServiceDecoratorTest`, I don't think it's necessary to check them again in the final test of `KubernetesJobManagerFactoryTest`, rough assert such as size verifying could be enough, otherwise, there are repeated test work.